File: web/src/contracts.ts
export type AnalysisMode = "policy" | "mcts";
export type PlayerSide = "white" | "black";

export interface AnalysisOptions {
  readonly mode: AnalysisMode;
  readonly time_limit_seconds: number;
}

export interface GameState {
  readonly starting_fen: string;
  readonly moves_uci: readonly string[];
  readonly fen: string;
  readonly side_to_move: PlayerSide;
  readonly game_over: boolean;
  readonly result: string | null;
}

export interface OutcomePrediction {
  readonly win: number;
  readonly draw: number;
  readonly loss: number;
  readonly perspective: "side_to_move";
}

export interface CandidateMove {
  readonly move_uci: string;
  readonly policy_prior: number;
  readonly visits: number;
  readonly visit_share: number;
  readonly mean_search_value: number | null;
}

export interface SearchMetrics {
  readonly searches: number;
  readonly maximum_depth: number;
  readonly elapsed_milliseconds: number;
}

export interface AnalysisResult {
  readonly chosen_move_uci: string;
  readonly root_value: number;
  readonly outcome_prediction: OutcomePrediction | null;
  readonly candidates: readonly CandidateMove[];
  readonly metrics: SearchMetrics;
  readonly principal_variation: readonly string[] | null;
}

export interface CreateGameResponse {
  readonly game_token: string;
  readonly state: GameState;
}

export interface PlayTurnResponse {
  readonly state: GameState;
  readonly engine_move_uci: string | null;
  readonly analysis: AnalysisResult | null;
}

export interface PlayTurnRequest {
  readonly starting_fen: string;
  readonly moves_uci: readonly string[];
  readonly human_move_uci: string | null;
  readonly analysis: AnalysisOptions;
}

File: py/web_play/contracts.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class FrozenModel(BaseModel):
    model_config = ConfigDict(frozen=True)


class AnalysisMode(str, Enum):
    POLICY = "policy"
    MCTS = "mcts"


class SideToMove(str, Enum):
    WHITE = "white"
    BLACK = "black"


class OutcomePerspective(str, Enum):
    SIDE_TO_MOVE = "side_to_move"


class OutcomePrediction(FrozenModel):
    win: float = Field(ge=0.0, le=1.0)
    draw: float = Field(ge=0.0, le=1.0)
    loss: float = Field(ge=0.0, le=1.0)
    perspective: Literal[OutcomePerspective.SIDE_TO_MOVE] = (
        OutcomePerspective.SIDE_TO_MOVE
    )


class CandidateMove(FrozenModel):
    move_uci: str
    policy_prior: float = Field(ge=0.0, le=1.0)
    visits: int = Field(ge=0)
    visit_share: float = Field(ge=0.0, le=1.0)
    mean_search_value: float | None = Field(default=None, ge=-1.0, le=1.0)


class SearchMetrics(FrozenModel):
    searches: int = Field(ge=0)
    maximum_depth: int = Field(ge=0)
    elapsed_milliseconds: int = Field(ge=0)


class AnalysisResult(FrozenModel):
    chosen_move_uci: str
    root_value: float = Field(ge=-1.0, le=1.0)
    outcome_prediction: OutcomePrediction | None
    candidates: tuple[CandidateMove, ...]
    metrics: SearchMetrics
    principal_variation: tuple[str, ...] | None


class AnalysisOptions(FrozenModel):
    mode: AnalysisMode
    time_limit_seconds: Annotated[int, Field(ge=1, le=30)]


class CreateGameRequest(FrozenModel):
    starting_fen: str
    moves_uci: Annotated[tuple[str, ...], Field(max_length=1024)]


class GameState(FrozenModel):
    starting_fen: str
    moves_uci: tuple[str, ...]
    fen: str
    side_to_move: SideToMove
    game_over: bool
    result: str | None


class CreateGameResponse(FrozenModel):
    game_token: UUID
    state: GameState


class PlayTurnRequest(FrozenModel):
    starting_fen: str
    moves_uci: Annotated[tuple[str, ...], Field(max_length=1024)]
    human_move_uci: str | None
    analysis: AnalysisOptions


class PlayTurnResponse(FrozenModel):
    state: GameState
    engine_move_uci: str | None
    analysis: AnalysisResult | None


@dataclass(frozen=True)
class TimedAnalysis:
    mode: AnalysisMode
    time_limit_seconds: int

    def __post_init__(self) -> None:
        if not 1 <= self.time_limit_seconds <= 30:
            raise ValueError("time_limit_seconds must be between 1 and 30.")


@dataclass(frozen=True)
class CountedAnalysis:
    mode: AnalysisMode
    searches: int

    def __post_init__(self) -> None:
        if self.searches < 1:
            raise ValueError("searches must be positive.")


AnalysisLimit = TimedAnalysis | CountedAnalysis
